go/batch-submitter/metrics/metrics.go

```diff
@@ -0,0 +1,79 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type Metrics struct {
+	// ETHBalance tracks the amount of ETH in the submitter's account.
+	ETHBalance prometheus.Gauge
+
+	// BatchSizeInBytes tracks the size of batch submission transactions.
+	BatchSizeInBytes prometheus.Histogram
+
+	// NumTxPerBatch tracks the number of L2 transactions in each batch
+	// submission.
+	NumTxPerBatch prometheus.Histogram
+
+	// SubmissionTimestamp tracks the time at which each batch was confirmed.
+	SubmissionTimestamp prometheus.Histogram
+
+	// SubmissionGasUsed tracks the amount of gas used to submit each batch.
+	SubmissionGasUsed prometheus.Histogram
+
+	// BatchsSubmitted tracks the total number of successful batch submissions.
+	BatchesSubmitted prometheus.Counter
+
+	// FailedSubmissions tracks the total number of failed batch submissions.
+	FailedSubmissions prometheus.Counter
+
+	// BatchTxBuildTime trakcs the duration it takes to construct a batch
+	// transaction.
+	BatchTxBuildTime prometheus.Gauge
+}
+
+func NewMetrics(subsystem string) *Metrics {
+	return &Metrics{
+		ETHBalance: promauto.NewGauge(prometheus.GaugeOpts{
+			Name:      "batch_submitter_eth_balance",
+			Help:      "ETH balance of the batch submitter",
+			Subsystem: subsystem,
+		}),
+		BatchSizeInBytes: promauto.NewHistogram(prometheus.HistogramOpts{
+			Name:      "batch_size_in_bytes",
+			Help:      "Size of batches in bytes",
+			Subsystem: subsystem,
+		}),
+		NumTxPerBatch: promauto.NewHistogram(prometheus.HistogramOpts{
+			Name:      "num_txs_per_batch",
+			Help:      "Number of transaction in each batch",
+			Subsystem: subsystem,
+		}),
+		SubmissionTimestamp: promauto.NewHistogram(prometheus.HistogramOpts{
+			Name:      "submission_timestamp",
+			Help:      "Timestamp of each batch submitter submission",
+			Subsystem: subsystem,
+		}),
+		SubmissionGasUsed: promauto.NewHistogram(prometheus.HistogramOpts{
+			Name:      "submission_gas_used",
+			Help:      "Gas used to submit each batch",
+			Subsystem: subsystem,
+		}),
+		BatchesSubmitted: promauto.NewCounter(prometheus.CounterOpts{
+			Name:      "batches_submitted",
+			Help:      "Count of batches submitted",
+			Subsystem: subsystem,
+		}),
+		FailedSubmissions: promauto.NewCounter(prometheus.CounterOpts{
+			Name:      "failed_submissions",
+			Help:      "Count of failed batch submissions",
+			Subsystem: subsystem,
+		}),
+		BatchTxBuildTime: promauto.NewGauge(prometheus.GaugeOpts{
+			Name:      "batch_tx_build_time",
+			Help:      "Time to construct batch transactions",
+			Subsystem: subsystem,
+		}),
```

Comment:
For new metrics, is there a "time waiting for confirmation" of some other way to track that the batch-submitter has submitted a tx and is waiting for it to be confirmed before starting the next one?

Reply:
Does that batch-submitter has an internal view of number of tx that need batching? Like how much work it has stacking up that we could monitor directly? (We currently get this indirectly by measuring the diff between sequencer and CTC totalElements).

Reply:
> or new metrics, is there a "time waiting for confirmation" of some other way to track that the batch-submitter has submitted a tx and is waiting for it to be confirmed before starting the next one?

Added a new `batch_submitter_batch_confirmation_time` metric!

> Does that batch-submitter has an internal view of number of tx that need batching? 

Currently no, as it greedily pulls in just enough L2 blocks to construct a full batch. Doing so would require fetching the entire backlog and, minimally, summing up the size of the pending transactions. I think for now we should continue with just observing the delta in height. In the future when we move to publishing concurrent batches, this will be effectively zero since the entire backlog can be processed instantly